packages/liveblocks-react/src/factory.tsx

```diff
@@ -804,7 +834,7 @@ Please see https://bit.ly/3Niy5aP for details.`
         unsubscribeRoot();
         unsubscribeCrdt();
       };
-    }, [root, room]);
+    }, [root, room, key, boxedInitialValue, rerender]);
```

Comment:
Omitting the key from this effect was intentional. I always thought about `useList`, `useMap`, `useObject` like `useState` where the arguments are only interpreted once (it made a bit more sense with the initializer).

It's still not clear to me how to express that a hook / component use some arguments / prop only once (for the initialization). Or maybe it's an anti pattern except for `useState` and `useRef`?

Reply:
I think it's fine to include it here, just in case people are using it dynamically? For example, in a chess game:

```tsx
const currentPlayer = useObject(turn % 2 === 0 ? 'whitePlayer' : 'blackPlayer');
const name = currentPlayer.get('name');
```

This would be fine I think? By including the `key` in here, we can just support this use case. If this code example would "stick" with `'whitePlayer'` on every turn, wouldn't that be unintuitive?

If we deliberately want to prevent changing the key, we should probably throw an error if the value ever differs from the initially received render. But I don't see why in this case. The deregister function will unsubscribe the old key, and subscribe to the new key. Seems nice to me!

I can see how this would have been weird in the past when there was an initial value 2nd argument, though.

I think not respecting changing inputs is an antipattern in React hooks. `useState` and `useRef` are exceptions because there you deliberately want to use the value as an initial value, but otherwise manage the state differently (by calling the setter, and by assigning to `.current`, respectively).

Reply:
> (I realize that could also be modeled differently.) But wouldn't this be just fine? By including the key in here, we can just support this use case. If this code example would "stick" with 'whitePlayer' on every turn, wouldn't that be unintuitive?

My message was not clear at all. I 100% agree with the fix you're proposing. Now that we removed the initializer, this behavior is the most intuitive.

I was just raising a more general question: how do we let users know that a hook argument or component prop is ignored after the first render? `RoomProvider` `initialStorage` and `initialPresence` almost have this behavior. 

I took a deeper look at new react doc, and I think there is an implicit convention that I like. Parameters that start with "initial" have this behavior: https://beta.reactjs.org/apis/usestate#parameters

Maybe this is a rule we can follow. 🤔

Reply:
Ahh, yes. I think that’s a good convention 👍

Reply:
Fixed in 1c6d7541d2c5f15ca439ad140cb0a8d592aecd0e.